### PR TITLE
encoding/wkt: replace wkt parser with goyacc parser

### LINKED ISSUE #151: encoding/wkt: wkt decoding case and space insensitive
Trying to replace my WKT decoding from cgo-ing with GEOS with the go-geom one here.

Unfortunately, it doesn't work for a couple of cases, e.g.
* `POINT(1.0 1.0)` should work (no space between `POINT` and `(`)
* `point (1.0 1.0)` should work (case insensitive POINT)
* `POINTZ (1.0 1.0)` should work (`Z` dimension / layout does not need a space afterwards)

Wondering if it's worth using a grammar file (https://godoc.org/golang.org/x/tools/cmd/goyacc) to do this instead.

I can work on this at a later date, but flagging as an issue for now. (It's not straightforward based on the way things are tokenised right now in the wkt decoder)

### LINKED ISSUE #180: WKT decoder could use re2c
[re2c](http://re2c.org/) just got a Go backend, and looks like a nice little lexer for WKT. Maybe we could use this to provide more robust WKT decoding?

### PR DESCRIPTION
**Pull request by andyyang890**
Fixes #151.
Fixes #180.

### CODE REVIEW

File: encoding/wkt/lex_errors.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	rightMax := e.lineStart + e.linePos + rightPadding
+	if snipEnd > rightMax {
+		snipEnd = rightMax
+		strLineSuffix = "..." + strLineSuffix
+	}
+
+	// Print a cursor pointing to the token where the problem occurred.
+	snippet := e.wkt[snipStart:snipEnd]
+	snippet = strings.ReplaceAll(snippet, "\t", " ")
+	err += strLinePrefix + snippet + strLineSuffix
```
Comment (andyyang890) [inline encoding/wkt/lex_errors.go:61]: @otan Made a minor change here compared to the version currently in the cockroach repo to handle the spacing with tabs

File: encoding/wkt/wkt.go [hunk 2]
```diff
@@ -57,5 +59,10 @@ func Marshal(g geom.T, applyOptFns ...EncodeOption) (string, error) {
 
 // Unmarshal translates a WKT to the corresponding geometry.
 func Unmarshal(wkt string) (geom.T, error) {
```
Comment (otan) [inline encoding/wkt/wkt.go:61]: i think in go-geom, we want to relax things we check in CockroachDB (subject to @twpayne guidance!):
a) geometrycollections can have mixed dimensions.
b) linestrings/polygons do not have have 2/4 points respectively, and polygon rings don't have to be closed.

we can choose to optionally allow these checks (but have them off by default). this should be done with options -- see how we did the `EncodeOption` for `Marshal` above.

Comment (andyyang890) [reply encoding/wkt/wkt.go:61]: > a) geometrycollections can have mixed dimensions.

Does this only apply to `GEOMETRYCOLLECTION` or does it apply to `GEOMETRYCOLLECTION{M, Z, ZM}` as well?

**Comment by twpayne** [comment]
Thank you @andyyang890 - this is a fantastic contribution!

I just back from a few days away but will review in detail in the next couple of days.

**Comment by twpayne** [comment]
Thank you so much for this @andyyang890!

The code is excellent, and thank you in particularly for the detail in handling edge cases.

I'll patch up the lint failure in a separate PR.

File: encoding/wkt/wkt.go (hunk 2 above)
Comment (twpayne) [reply encoding/wkt/wkt.go:61]: Very good points! I've opened #200 to continue the discussion.

**Comment by andyyang890** [comment]
Sounds good, thanks for the detailed review!

**Pull request merged**